src/lib/pubkey/kyber/kyber/kyber_modern.h

```diff
@@ -24,44 +24,75 @@ namespace Botan {
 class Kyber_Modern_Symmetric_Primitives : public Kyber_Symmetric_Primitives
    {
    public:
+      Kyber_Modern_Symmetric_Primitives() :
+         m_sha3_512(HashFunction::create_or_throw("SHA-3(512)")),
+         m_sha3_256(HashFunction::create_or_throw("SHA-3(256)")),
+         m_shake256_256(HashFunction::create_or_throw("SHAKE-256(256)"))
+         {}
+
       std::unique_ptr<HashFunction> G() const override
          {
-         return HashFunction::create_or_throw("SHA-3(512)");
+         return m_sha3_512->new_object();
          }
 
       std::unique_ptr<HashFunction> H() const override
          {
-         return HashFunction::create_or_throw("SHA-3(256)");
+         return m_sha3_256->new_object();
          }
 
       std::unique_ptr<HashFunction> KDF() const override
          {
-         return HashFunction::create_or_throw("SHAKE-256(256)");
+         return m_shake256_256->new_object();
          }
```

Comment:
Interesting idiom. I guess the benefit comes from the avoided algorithm lookup. :+1:

Reply:
Kyber is so fast that *any* overhead makes a big difference in overall performance. Ordinarily this approach would not be worth the bother.